### ISSUE TITLE
Scenes handling is perhaps buggy?

### ISSUE THREAD

**Issue by fterh** [telegraf/telegraf#369]
I tested with [scenes-bot.js](https://github.com/telegraf/telegraf/blob/develop/docs/examples/scenes-bot.js), and here are my observations:

1. Sometimes, I have to enter the echo scene twice for it to actually start echoing. The first time I run /echo, it replies with "echo scene" but does not actually start echoing (instead, it returns the "Try /echo or /greeter" message).

2. When I send /back, sometimes, I do not get "exiting echo scene". It just leaves scene, because my next message is met with the "Try /echo or /greeter" again.

What's really frustrating is that my observations are not consistent.

**Comment by roccomuso** [comment]
I'm having similar issues.
Having a default scene in my stage middleware and sometimes my enter callback is not called.

```javascript
greeterScene.enter((ctx) => ctx.reply('Hi'))
```

**Comment by ImTheDeveloper** [comment]
Did you guys ever get this working? Also hitting the same issue if I programmaticly leave one scene and enter another. It seems to enter fine but then doesn't pick up anything from ``` .on('text'```

**Comment by TorlanDelta** [comment]
Many different issues with scenes. 
Sometimes `enter` just refuses to work. It just doesn't. Especially when using Composers with scenes. This is just beyond me.

**Comment by TorlanDelta** [comment]
Okay, people. This is what I found out. 

I was playing around with this [example](https://github.com/telegraf/telegraf/blob/develop/docs/examples/scenes-bot.js) and I have finally found the possible cause of most of the problems. 

handlers like 
```
bot.command('greeter', enter('greeter'))
```
always work, HOWEVER
```
bot.command('greeter', ctx => {
   enter('greeter')
})
```
never do. Found it out by myself. I used to create handlers for entering one scene from within another, trying to save something to the `ctx.session` object first like so - 

```
scene1.action(/^thang$/, ctx => {
   ctx.session.thang = ctx.callbackQuery.data // ='thang'
   enter('scene2')
}) 

scene2.enter( ctx => {
    //do stuff
})
```
But the correct(and quite unclearly so, especially because of this [example](https://github.com/telegraf/telegraf/blob/develop/docs/examples/self-cleaning-scenes.js) - notice the `async ctx => enter('stuff') )   way would be: 
```
scene1.action(/^thang$/, enter('scene2')) 

scene2.enter( ctx => {
   ctx.session.thang = ctx.callbackQuery.data //still  ='thang' !!!
    //do stuff
})
```
Four-five hours well spent, huh
